15692: Introduce background jobs

alehaa opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note that
    our contribution policy requires that a feature request or bug report be
    approved and assigned prior to opening a pull request. This helps avoid
    waste time and effort on a proposed change that we might not be able to
    accept.

    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ISSUE WHICH HAS BEEN ASSIGNED
    TO YOU, IT WILL BE CLOSED AUTOMATICALLY.

    Please specify your assigned issue number on the line below.
-->
### Fixes: #15692 

<!--
    Please include a summary of the proposed changes below.
-->

This PR includes three new background job class to be used in NetBox and by NetBox plugins. They handle the necessary logic for starting and stopping jobs, including exception handling and rescheduling of recurring jobs.

This PR also implements the use of this framework for existing data source and script jobs, eliminating duplicate code and ensuring consistency.

alehaa committed:
Introduce reusable BackgroundJob framework

A new abstract class can be used to implement job function classes. It
handles the necessary logic for starting and stopping jobs, including
exception handling and rescheduling of recurring jobs.

This commit also includes the migration of data source jobs to the new
framework.

alehaa committed:
Restore using import_string for jobs

Using the 'import_string()' utility from Django allows the job script
class to be simplified, as module imports no longer need to avoid loops.
This should make it easier to queue and maintain jobs.

alehaa committed:
Use SyncDataSourceJob for management command

Instead of maintaining two separate job execution logics, the same job
is now used for both background and interactive execution.

alehaa committed:
Implement BackgroundJob for running scripts

The independent implementations of interactive and background script
execution have been merged into a single BackgroundJob implementation.

alehaa committed:
Fix documentation of model features

alehaa committed:
Ensure consitent code style

alehaa committed:
Introduce reusable ScheduledJob

A new abstract class can be used to implement job function classes that
specialize in scheduling. These use the same logic as regular
BackgroundJobs, but ensure that they are only scheduled once at any given
time.

alehaa committed:
Introduce reusable SystemJob

A new abstract class can be used to implement job function classes that
specialize in system background tasks (e.g. synchronization or
housekeeping). In addition to the features of the BackgroundJob and
ScheduledJob classes, these implement additional logic to not need to be
bound to an existing NetBox object and to setup job schedules on plugin
load instead of an interactive request.

alehaa committed:
Add documentation for jobs framework

alehaa committed:
Merge branch 'feature' into 15692-background-jobs

jeremystretch commented on netbox/core/management/commands/syncdatasource.py:
```diff
@@ -34,13 +34,12 @@ def handle(self, *args, **options):
-                DataSource.objects.filter(pk=datasource.pk).update(status=DataSourceStatusChoices.FAILED)
-                raise e
+
+            datasource.enqueue_sync_job()
```
IMO we should leave this as-is and avoid the use of a background job entirely. We want the management command to block on the call to `sync()`.

alehaa replied:
When `enqueue_sync_job()` gets no request passed as argument, the job will be run in foreground and block until `sync()` has finished. See netbox/core/models/data.py line 169. Maybe a dedicated `background=False` parameter would make this more clear than just using some automation?

jeremystretch replied:
Right, but there's no reason to enqueue a job at all; the management command intends to call `sync()` directly. This change can be reverted.

jeremystretch commented on netbox/core/models/jobs.py:
```diff
@@ -224,6 +226,12 @@ def enqueue(cls, func, instance, name='', user=None, schedule_at=None, interval=
         )
 
+        # Optionally, the job can be run immediately without being scheduled to run in the background.
+        if run_now:
+            func(job_id=str(job.job_id), job=job, **kwargs)
+            return job
```
I could be missing something, but this seems like it defeats the purpose of using a job. Or if the goal is to reuse logic defined within a job, could we just call its `run()` method directly?

alehaa replied:
My intention was to run a job in the foreground, but also create a `Job` object for documentation. Otherwise, jobs would run on the console, but there would be no audit trail. Now there's at least the information that a job was run, although the changelog objects are missing (maybe a topic for a future FR).

I agree that the method name `enqueue()` is a bit misleading. I could rename it to `run()` with the default parameter `background=True` to run jobs in the background, but also give the option to run it now. Since the `Job` API is not yet public for plugins, this should not break compatibility.

jeremystretch commented on netbox/utilities/jobs.py:
```diff
@@ -0,0 +1,196 @@
+        return Job.enqueue(cls.handle, *args, **kwargs)
+
+
+class ScheduledJob(BackgroundJob):
```
Could we just collapse the functionality delivered by this class into BackgroundJob by adding an alternative `enqueue_once()` method to it? This would replicate the functionality of the `schedule()` method without requiring a separate class.

alehaa replied:
I was hesitant at first because `ScheduledJob` uses a slightly different logic by hardcoding job names. However, I think it can be done and I give it a try.

jeremystretch commented on netbox/utilities/jobs.py:
```diff
@@ -0,0 +1,196 @@
+        return cls.enqueue(instance=instance, interval=interval, *args, **kwargs)
+
+
+class SystemJob(ScheduledJob):
```
Would it remove the need for this class if we allowed a Job to be enqueued without defining an object? I don't actually think it was intended to be required: The Job model currently requires an `object_type` but not an `object_id`. I'm open to removing that requirement if it simplifies this effort.

alehaa replied:
Yep. The sole purpose of this class is to circumvent the required `instance` parameter of `Job.enqueue()` and adding the `setup()` method. I'll add a commit to make the `instance` parameter optional and thus removing the need for this class.

jeremystretch reviewed (commented):
@alehaa thank you for such a thorough and well-documented PR! Apologies again for the delay in review. At a high level, this all looks doable. I just have a few questions about the implementation details.

alehaa commented on netbox/utilities/jobs.py:
```diff
@@ -0,0 +1,196 @@
+    method is called.
+    """
+
+    ENQUEUED_STATUS = [
```
Should be moved into netbox/netbox/core/choices.py

alehaa commented on netbox/utilities/jobs.py:
```diff
@@ -0,0 +1,196 @@
+    method is called.
+    """
+
+    ENQUEUED_STATUS = [
+        JobStatusChoices.STATUS_PENDING,
+        JobStatusChoices.STATUS_SCHEDULED,
+        JobStatusChoices.STATUS_RUNNING,
+    ]
```
Should be moved into netbox/netbox/core/choices.py

alehaa committed:
Revert "Use SyncDataSourceJob for management"

This partially reverts commit db591d4. The 'run_now' parameter of
'enqueue()' remains, as its being used by following commits.

alehaa committed:
Merge enqueued status into JobStatusChoices

alehaa committed:
Fix logger for ScriptJob

alehaa committed:
Remove job name for scripts

Because scripts are already linked through the Job Instance field, the
name is displayed twice. Removing this reduces redundancy and opens up
the possibility of simplifying the BackgroundJob framework in future
commits.

alehaa committed:
Merge ScheduledJob into BackgroundJob

Instead of using separate classes, the logic of ScheduledJob is now
merged into the generic BackgroundJob class. This allows reusing the
same logic, but dynamically deciding whether to enqueue the same job
once or multiple times.

alehaa committed:
Add name attribute for BackgroundJob

Instead of defining individual names on enqueue, BackgroundJob classes
can now set a job name in their meta class. This is equivalent to other
Django classes and NetBox scripts.

jeremystretch committed:
Drop enqueue_sync_job() method from DataSource

jeremystretch committed:
Import ScriptJob directly

jeremystretch committed:
Relax requirement for Jobs to reference a specific object

jeremystretch committed:
Rename 'run_now' arg on Job.enqueue() to 'immediate'

alehaa commented:
@jeremystretch I‘m a bit confused about your commits. You requested changes, I‘ve worked a day on them and just before uploading commits apear in this PR. 🧐

jeremystretch commented:
@alehaa apologies for the confusion; it had been several days with no response so I assumed you did not have time to make the changes. I've undertaken the work we discussed above as well as some general cleanup. We need to have all the v4.1 work merged by end of week.

jeremystretch commented:
@alehaa do you want to handle renaming ScheduledJob from here, or should I just wrap it up?

alehaa commented:
@jeremystretch no worries, but I’d like to keep it up from here. I see some of your code handles parts I‘ve not finished yet. Please give me till tomorrow to integrate your changes into mine and I‘ll upload the combination from both of our sources.

alehaa committed:
Merge branch '15692-cherry' into 15692-background-jobs

Unfortunately, work on this branch was done twice. However, most of the
commits affected different sections of the code. This merge commit
combines both efforts, but removes duplicate work.

alehaa committed:
Fix queue lookup in Job enqueue

alehaa committed:
Collapse SystemJob into BackgroundJob

alehaa committed:
Remove legacy JobResultStatusChoices

ChoiceSet was moved to core in 40572b5.

alehaa committed:
Use queue 'low' for system jobs by default

System jobs usually perform low-priority background tasks and therefore
can use a different queue than 'default', which is used for regular jobs
related to specific objects.

alehaa committed:
Add test cases for BackgroundJob handling

alehaa committed:
Fix enqueue interval jobs

As the job's name is set by enqueue(), it must not be passed in handle()
to avoid duplicate kwargs with the same name.

alehaa committed:
Honor schedule_at for job's enqueue_once

Not only can a job's interval change, but so can the time at which it is
scheduled to run. If a specific scheduled time is set, it will also be
checked against the current job schedule. If there are any changes, the
job is rescheduled with the new time.

alehaa committed:
Switch BackgroundJob to regular methods

Instead of using a class method for run(), a regular method is used for
this purpose. This gives the possibility to add more convenience methods
in the future, e.g. for interacting with the job object or for logging,
as implemented for scripts.

alehaa commented on netbox/utilities/jobs.py:
```diff
@@ -0,0 +1,145 @@
+        This method should be called from the plugins `ready()` function to set up the schedule as early as possible.
+        For interactive setup of schedules (e.g. on user requests), use either `enqueue()` or `enqueue_once()` instead.
+        """
+        connection_created.connect(lambda sender, **signal_kwargs: cls.enqueue_once(*args, **kwargs))
```
A minor problem with using the `connection_created' signal is that it's fired on every database connection. This includes every time a job is run by the rqworker. In general this should not have any side effects, but performance will be slightly degraded as the schedules of system background jobs will be checked every time.

To improve performance, I suggest adding an exit-early if `setup()` is not called by a WSGI process, and perhaps only for the first thread. That way, only a subset of NetBox processes will handle the setup of system background jobs. However, I'm not familiar enough with NetBox's WSGI internals to know how to detect these conditions.

jeremystretch replied:
I may not be clear on the exact functionality here. Would it suffice to just call `enqueue_once()` directly from the AppConfig's `ready()` method, rather than attaching it to a signal?

The `connection_created` signal isn't really tenable for this, as database connections can be created quite frequently depending on the setup, especially when a large number of WSGI workers are in use.

jeremystretch replied:
True, this is a bit tricky. It feels like there should be a canonical convention for this, but I'm not sure what it would be.

Unfortunately we're on a clock here, as the first v4.1 beta release is due August 5th, and I'd rather not hold up the PR just for this issue. How would you feel about splitting out the auto-run functionality for now? We can merge the remainder of the PR largely as-is (it probably deserves one final review), and revisit the auto-run piece individually. If we arrive at a workable solution, I'd be comfortable slipping that in during the beta period ahead of the final release.

alehaa replied:
That's fine by me. I've removed the `setup()` method and documentation for system jobs. I think we can close this FR without auto-registration and solve this task with #16971.

The code is ready to be used as a system background job, only the auto-registration isn't solved yet. Although not recommended, anyone can already call `enqueue_once()` in the `ready()` method of AppConfig and it *should* work (no guarantee).

alehaa committed:
Fix background tasks documentation

alehaa committed:
Test enqueue in combination with enqueue_once

alehaa commented:
@jeremystretch I've finished my work and added everything to this PR and its ready for review now. Please note that I did need to do a force push, as some of our works did conflict and I didn't want to open a third PR for this issue. Therefore you might need to to a `git fetch` followed by a `git reset --hard origin/15692-background-jobs` to get the current state on your local machine.

Despite the open tasks and some bug fixing, I've 
* integrated additional tests for the `BackgroundJob` framework
* did some cleanup of legacy code in 309ad29
* implemented to use queue `low` for system background jobs by default
* switched `run()` from a `@classmethod` to a regular method to simplify implementation and give options for possible future extensions

Also note the open comment on the `setup()` method above, maybe you can take a look at it? Feel free to add or request changes to the code as you see fit.

jeremystretch committed:
Rename background jobs to tasks (to differentiate from RQ)

jeremystretch committed:
Touch up docs

jeremystretch commented on netbox/netbox/settings.py:
```diff
@@ -124,7 +124,9 @@
 PLUGINS = getattr(configuration, 'PLUGINS', [])
 PLUGINS_CONFIG = getattr(configuration, 'PLUGINS_CONFIG', {})
-QUEUE_MAPPINGS = getattr(configuration, 'QUEUE_MAPPINGS', {})
+QUEUE_MAPPINGS = getattr(configuration, 'QUEUE_MAPPINGS', {
+    None: 'low',
+})
```
We should leave this as-is, because it would effect a change to the default configuration, which would require a separate FR. I assume this is for system jobs? If so, we could instead explicitly queue the job in `RQ_QUEUE_LOW`.

alehaa replied:
I've reverted b17b2050df953427769ab6a337c6995ac2a60615. For specifying a custom queue I would not hardcode the low queue, but wait for #16516 to be resolved. As all enqueue methods of `BackgroundJob` pass through `kwargs`, a future `queue` parameter (as suggested in this FR) will be passed through as well.

alehaa committed:
Revert "Use queue 'low' for system jobs by default"

This reverts commit b17b2050df953427769ab6a337c6995ac2a60615.

alehaa committed:
Remove system background job

This commit reverts commits 4880d81 and 0b15ecf. Using the database
'connection_created' signal for job registration feels a little wrong at
this point, as it would trigger registration very often. However, the
background job framework is prepared for this use case and can be used
by plugins once the auto-registration of jobs is solved.

arthanson commented on netbox/core/models/jobs.py:
```diff
@@ -197,25 +199,34 @@ def terminate(self, status=JobStatusChoices.STATUS_COMPLETED, error=None):
 
     @classmethod
-    def enqueue(cls, func, instance, name='', user=None, schedule_at=None, interval=None, **kwargs):
+    def enqueue(cls, func, instance=None, name='', user=None, schedule_at=None, interval=None, immediate=False, **kwargs):
```
I think it might be cleaner to break this into a base function with two other functions enqueue and run? to separate out the immediate keyword and schedule_at / interval.   In the immediate case you are not really enqueueing anything. I'm not sure interval is supported when using Immediate?  It is not used by the management command.

alehaa replied:
There was an earlier discussion about naming the parameter and the consequences. You could split this into two methods, both reusing a third internal `_create()` method to make the parameters and behavior a bit clearer to the user. This would also require an additional proxy method in `BackgroundJob`. I'd like to leave the decision to @jeremystretch.

arthanson commented on netbox/utilities/tests/test_jobs.py:
```diff
@@ -0,0 +1,134 @@
+        self.assertRaises(Job.DoesNotExist, job1.refresh_from_db)
+        self.assertEqual(TestBackgroundJob.get_jobs(instance).count(), 1)
+
+    def test_enqueue_system(self):
```
Would be good to add a test with the name parameter.

alehaa replied:
@arthanson can you elaborate on this please? Passing `name` in `enqueue()` and `enqueue_once()` is not supported as documented. Test for automatic setting of the job's `name` are implemented in `BackgroundJobTest`.

arthanson replied:
Partly it was used in runscript but I guess that might have been a mistake as it looks like it is removed now, but I'm not understanding why it is a parameter if it can't be passed in?

alehaa replied:
One can pass a `name` to a `Job` object when using it standalone. `BackgroundJob` augments `Job` and sets the `name` parameter automatically, so it must not be set when calling `BackgroundJob.enqueue()`.

I forgot removing the parameter in `runscript`. This was fixed with ecf8e79fddcfd5b885fac3c29231edbcd73a12d8.

arthanson commented on netbox/core/models/jobs.py:
```diff
@@ -197,25 +199,34 @@ def terminate(self, status=JobStatusChoices.STATUS_COMPLETED, error=None):
-        object_type = ObjectType.objects.get_for_model(instance, for_concrete_model=False)
-        rq_queue_name = get_queue_for_model(object_type.model)
+        if schedule_at and immediate:
+            raise ValueError("enqueue() cannot be called with values for both schedule_at and immediate.")
```
Should be internationalized with _(

alehaa committed:
Fix runscript management command

Defining names for background jobs was disabled with fb75389. The
preceeding changes in 257976d did forget the management command.

alehaa commented on netbox/extras/management/commands/runscript.py:
```diff
@@ -120,40 +76,31 @@ def _run_script():
+
+        # Execute the script.
+        ScriptJob = import_string("extras.jobs.ScriptJob")
+        job = ScriptJob.enqueue(
```
Seems I forgot one call with this parameter in 257976d815605206686ebbf1c43db149131fe0fe. Fixed with ecf8e79fddcfd5b885fac3c29231edbcd73a12d8.

alehaa committed:
Use regular imports for ScriptJob

jeremystretch commented on netbox/extras/api/views.py:
```diff
@@ -273,10 +273,9 @@ def post(self, request, pk):
         if input_serializer.is_valid():
-            Job.enqueue(
-                run_script,
+            ScriptJob = import_string("extras.jobs.ScriptJob")
```
I think this is needed to avoid a circular import, however it should be fine to import `ScriptJob` directly (`from extras.jobs import ScriptJob`) at this point.

alehaa replied:
This is a relict from using `BackgroundJob` for syncing data sources to avoid circular imports. For `ScriptJob` this is not necessary, so I've changed as requested in 7f0a4e3c34c06891d225535c36e3389fa6a8705d.

arthanson commented:
BTW @alehaa IMHO excellent work on this, both the writeup in the issue and the PR itself is very high quality and impressive.  Nice work!

jeremystretch commented:
Seems like we're just about there, however I still have one concern: As it stands, I expect the average user will have difficulty distinguishing between the Job database model and the BackgrounJob runner class. I'd like to rename `BackgroundJob` to `JobRunner` for clarity. IMO this better conveys the relationship between classes which embody the programming logic and the database records which track its execution.

jeremystretch committed:
Rename BackgroundJob to JobRunner

jeremystretch commented:
Thanks again for all your work on this @alehaa!
